File: src/Creational_Patterns/Singleton/examples/go/singleton.go
package main

import (
	"fmt"
	"sync"
)

type Database struct {
	// Some fields and methods
	// ...
}

var (
	instance *Database
	once     sync.Once
)

func GetInstance() *Database {
	once.Do(func() {
		instance = &Database{}
		// Initialize the database instance
	})
	return instance
}

func (db *Database) Query(query string) {
	// Perform the database query
	// ...
}

func main() {
	// Access the singleton instance
	db := GetInstance()

	// Use the singleton instance
	db.Query("SELECT * FROM users;")
}

File: src/Creational_Patterns/Singleton/examples/rust/singleton.rs
// Cargo.toml ==>
// [dependencies]
//lazy_static = "1.4"

#[macro_use]
extern crate lazy_static;

use std::sync::Mutex;

lazy_static! {
    static ref DATABASE: Mutex<Database> = Mutex::new(Database::new());
}

struct Database {
    // Some fields and methods
    // ...
}

impl Database {
    fn new() -> Self {
        // Initialize the database instance
        Database {
            // ...
        }
    }

    fn query(&self, query: &str) {
        // Perform the database query
        // ...
    }
}

fn main() {
    // Access the singleton instance
    let db = DATABASE.lock().unwrap();

    // Use the singleton instance
    db.query("SELECT * FROM users;");
}